File: apps/ios/Packages/ShidouKit/Sources/ShidouProtocol/Herdr.swift
import Foundation

public enum HerdrAgentStatus: String, Codable, Hashable, Sendable {
    case idle, working, blocked, done, unknown
}

public struct HerdrAgentSession: Codable, Hashable, Sendable {
    public var source: String
    public var agent: String
    public var kind: String
    public var value: String
}

public struct HerdrWorktree: Codable, Hashable, Sendable {
    public var repoName: String
    public var repoRoot: String
    public var checkoutPath: String
    public var isLinkedWorktree: Bool

    enum CodingKeys: String, CodingKey {
        case repoName = "repo_name"
        case repoRoot = "repo_root"
        case checkoutPath = "checkout_path"
        case isLinkedWorktree = "is_linked_worktree"
    }
}

public struct HerdrWorkspace: Codable, Identifiable, Hashable, Sendable {
    public var id: String
    public var label: String
    public var focused: Bool
    public var paneCount: Int
    public var tabCount: Int
    public var status: HerdrAgentStatus
    public var worktree: HerdrWorktree?

    enum CodingKeys: String, CodingKey {
        case id, label, focused, status, worktree
        case paneCount = "pane_count"
        case tabCount = "tab_count"
    }
}

public struct HerdrAgent: Codable, Identifiable, Hashable, Sendable {
    public var paneId: String
    public var terminalId: String
    public var workspaceId: String
    public var tabId: String
    public var agent: String
    public var name: String?
    public var title: String?
    public var cwd: String?
    public var status: HerdrAgentStatus
    public var focused: Bool
    public var revision: UInt64
    public var providerSession: HerdrAgentSession?

    public var id: String { terminalId }
    public var displayTitle: String { title ?? name ?? agent }

    enum CodingKeys: String, CodingKey {
        case agent, name, title, cwd, status, focused, revision
        case paneId = "pane_id"
        case terminalId = "terminal_id"
        case workspaceId = "workspace_id"
        case tabId = "tab_id"
        case providerSession = "provider_session"
    }
}

public struct HerdrState: Codable, Equatable, Sendable {
    public var available: Bool
    public var version: String?
    public var `protocol`: UInt32?
    public var workspaces: [HerdrWorkspace]
    public var agents: [HerdrAgent]
    public var unavailableReason: String?

    enum CodingKeys: String, CodingKey {
        case available, version, `protocol`, workspaces, agents
        case unavailableReason = "unavailable_reason"
    }

    public init(
        available: Bool,
        version: String?,
        protocol: UInt32?,
        workspaces: [HerdrWorkspace],
        agents: [HerdrAgent],
        unavailableReason: String?
    ) {
        self.available = available
        self.version = version
        self.protocol = `protocol`
        self.workspaces = workspaces
        self.agents = agents
        self.unavailableReason = unavailableReason
    }
}

public struct HerdrAgentOutput: Codable, Equatable, Sendable {
    public var paneId: String
    public var text: String
    public var revision: UInt64
    public var truncated: Bool

    enum CodingKeys: String, CodingKey {
        case text, revision, truncated
        case paneId = "pane_id"
    }
}

File: crates/shidou-protocol/src/herdr.rs
//! Herdr resources exposed through the Shidou daemon.
//!
//! Herdr owns terminal workspaces and agent processes. These types preserve
//! that model instead of forcing a terminal-backed agent into `AgentSession`,
//! whose transcript and turn guarantees do not apply.

use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Serialize, TS)]
#[serde(rename_all = "camelCase")]
pub enum HerdrAgentStatus {
    Idle,
    Working,
    Blocked,
    Done,
    #[default]
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, TS)]
pub struct HerdrAgentSession {
    pub source: String,
    pub agent: String,
    pub kind: String,
    pub value: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, TS)]
pub struct HerdrWorktree {
    pub repo_name: String,
    pub repo_root: String,
    pub checkout_path: String,
    pub is_linked_worktree: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, TS)]
pub struct HerdrWorkspace {
    pub id: String,
    pub label: String,
    pub focused: bool,
    pub pane_count: usize,
    pub tab_count: usize,
    pub status: HerdrAgentStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub worktree: Option<HerdrWorktree>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, TS)]
pub struct HerdrAgent {
    /// Herdr's public pane id. Refresh it from each snapshot because moving a
    /// pane across workspaces can change this id.
    pub pane_id: String,
    /// Stable identity for the live terminal while its pane moves.
    pub terminal_id: String,
    pub workspace_id: String,
    pub tab_id: String,
    pub agent: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    pub status: HerdrAgentStatus,
    pub focused: bool,
    pub revision: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider_session: Option<HerdrAgentSession>,
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize, TS)]
pub struct HerdrState {
    pub available: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub protocol: Option<u32>,
    #[serde(default)]
    pub workspaces: Vec<HerdrWorkspace>,
    #[serde(default)]
    pub agents: Vec<HerdrAgent>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unavailable_reason: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, TS)]
pub struct HerdrAgentOutput {
    pub pane_id: String,
    pub text: String,
    pub revision: u64,
    pub truncated: bool,
}
